rootfs/apply.go

```diff
@@ -0,0 +1,136 @@
+package rootfs
+
+import (
+	"io"
+	"io/ioutil"
+
+	"github.com/docker/containerd"
+	"github.com/docker/containerd/log"
+	"github.com/docker/distribution"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+)
+
+type Snapshotter interface {
+	Prepare(key, parent string) ([]containerd.Mount, error)
```

Comment:
@samuelkarp Yes, this model has been validated across the graphdrivers. We are trying to avoid the situation where the graphdrivers are vertically integrated with the entire system. This makes container runtimes a lot more flexible and opens up more distribution models.

This pseudo-implementation provides the full facilities for de-duplicating layers, taking into account compressed and uncompressed hashes, as well as chained application of layers. You can read [here](https://github.com/docker/containerd/pull/427/files#diff-6ea9abd57ac207460460749d57c0b8f9R99) to better understand that. There may be more metadata relationships for the layers at the top-level, but we'll level that to the caller to decide, rather than trying to do too much in containerd. Again, the key is to avoid the vertically integrated mess of graphdrivers.